Issue #1513 Document how to write child device config management agent (#1536)
thin-edge/thin-edge.io: docs/src/howto-guides/child_device_config_managament.md

These requests arrive in the following JSON format:

```json
{
    "type": "{config-type}",
    "path": "/child/local/fs/path",
    "url": "http://{tedge-ip}:8000/tedge/file-transfer/{child-d}/config_snapshot/{config-type}"
}
```

The `type` and `path` fields are the same values that the child device sent to thin-edge in its `c8y-configuration-plugin.toml` file.
The `url` value is what the child device agent must use to upload the requested config file.

On receipt of the request, the agent must send an "executing" MQTT status message as follows:

**Topic:**

`tedge/{child-d}/commands/res/config_snapshot`

**Payload**:

```json
{
    "status": "executing",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

After sending this status message, the agent must upload the requested configuration file content to
the `url` received in the request with an HTTP PUT request.

Once the upload is complete, send a "successful" MQTT status message as follows:

**Topic:**

`tedge/{child-d}/commands/res/config_snapshot`

**Payload**:

```json
{
    "status": "successful",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

> If there are any failures while reading or uploading the requested conifg file,

Reviewer: ```suggestion
If there are any failures while reading or uploading the requested config file,
```

a "failed" status update must be sent instead, to the same topic as follows:

```json
{
    "status": "failed",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

# Handle config update requests from thin-edge

Handling config update requests from thin-edge is a 5-step process:

> 1. Subsribe to, and receive conifg update requests via MQTT

Reviewer: ```suggestion
1. Subscribe to, and receive config update requests via MQTT
```

}
```

The child device agent can download the conifg file update from thin-edge using the `url` for the given `type`.

On receipt of the request, the agent must send an "executing" MQTT status message as follows:

**Topic:**

`tedge/{child-d}/commands/res/config_update`

**Payload**:

```json
{
    "status": "executing",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

After sending this status message, the agent must download the configuration file update
from the `url` received in the request with an HTTP GET request.
The agent can then apply the downloaded configuration file update on the device.

Once the update is applied, send a "successful" MQTT status message as follows:

**Topic:**

`tedge/{child-d}/commands/res/config_update`

**Payload**:

```json
{
    "status": "successful",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

If there are any failures while downloading and applying the update,
a "failed" status update must be sent instead, to the same topic as follows:

```json
{
    "status": "failed",
    "type": "{config-type}",
    "path": "/child/local/fs/path" 
}
```

> ## Reference Implementation

Reviewer: I would also add a link to the reference documentation.

```suggestion
## References

 * [documentation](...)
 * [implementation example](...)
```

> [Here](https://github.com/thin-edge/thin-edge.io_spikes/blob/main/file-transfer-service/child_device.py) is a reference child device agent implementaion written in Python to demonstrate the above mentioned protocol.

Reviewer: This example must move to the [example repository](https://github.com/thin-edge/thin-edge.io_examples),
the spike repository being private.

Author: Done in this PR: https://github.com/thin-edge/thin-edge.io_examples/pull/36
